doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+    nexthop_group = NEXT_HOP_GROUP_TABLE:key ; index within the NEXT_HOP_GROUP_TABLE, optionally used instead of nexthop and intf fields
+ ```
+
+## 3.3 Switch State Service Design
+### 3.3.1 Orchestration Agent
+
+A new orchestration agent will be written to handle the new NEXT_HOP_GROUP_TABLE in APP_DB. For a new or updated entry in the NEXT_HOP_GROUP_TABLE, programming will depend on whether the group is configured with 1 or multiple next hops.
+
+ - If the group has a single next hop, the next hop group orchagent will simply get the SAI identifier for that next hop from the neighbor orchagent, and use that as the SAI identifier for the group.
+ - If the group has multiple next hops, the next hop group orchagent will add a next hop group to ASIC_DB and then add a next hop group member to ASIC_DB for every member of that group that is available to be used. Changes to the membership of the group will result in next hop group members being added to or removed from ASIC_DB. The next hop group orchagent will then maintain an association between the identifer of the group from APP_DB and the SAI identifier assigned to the next hop group.
+
+The handling of whether a next hop group member can be programmed or not, interfaces going up and down, etc. will match the same handling done by next hop groups managed by the route orchagent.
+
+When the next hop group is deleted, the next hop group orchagent will remove the next hop group from the ASIC_DB if required, and remove the association between the group identifier from APP_DB and the SAI identifier from ASIC_DB.  This will only happen once the next hop group is no longer referenced by any routes - and so the next hop group will maintain a reference count to ensure this.
+
+The route orchagent will parse the new nexthop_group field, and call into the next hop group orchagent to find the corresponding SAI object ID to program as the next hop attribute.  If the next hop group does not exist, the route will remain on the list of routes to sync, and will be retried on future runs of the route orchagent.
+
+If a next hop group cannot be programmed because the data plane limit has been reached, one next hop will be picked to be temporarily used for that group.  When a route is then programmed using that next hop group, it will be informed that a temporary form of the next hop group is in use, and so the route will remain on the pending list and will retry until the correct next hop group is programmed.  This mirrors the current behaviour in the route orchagent when the next hop group limit is reached.
```

Comment:
>it will be informed [](start = 215, length = 19)

In this situation, what does the route entry and next hop group entry look like?

Reply:
From the discussion I understood that when the route changes from ECPM to single path or vice-versa, the application ( frr-zebra) needs to reprogram the route so that NH identifier will be updated correctly. ? Is this understanding correct ? if so, does frr-zebra support this behaviour already ?

Reply:
Currently frr-zebra will reprogram the route whenever its next hops change at all. If it were to change to use this new schema (which is not in scope at this time) it would need to ensure that it still updated the route when changing between single and multiple next hops

Reply:
Thanks for the reply.  Will it be better to handle the reprogramming at orchagent level, since that is where we are differentiating between the ECMP and Non-ECPMP NH  type ?

Reply:
It would be possible, but would require orchagent to keep track of which routes are using each next hop/next hop group, which would add a lot of storage. This is because changing between the two will change the SAI object ID, and so every route using the next hop/group would need to be updated. I would suggest keeping the current requirement for now, and removing it in future if there is a clear need to do so.

Reply:
To clarify my original question:
In this situation, what does the route entry and next hop group entry look like in AppDB?

Reply:
They remain the same